File: src/qa/bfla.ts
// ── QA Touchstone — function-level authz (BFLA) engine (pure logic) ──────────
// OWASP API5: a lower-privilege identity must not be able to invoke a higher-
// privilege FUNCTION. The full matrix can express this cell-by-cell; this engine
// auto-derives the high-value subset — every privileged endpoint × every
// non-privileged identity, expectation deny — so a user gets a one-click BFLA
// scan without hand-building a matrix. It composes the shared authz primitives
// (privileged detection + soft-deny-aware classification), it does not re-derive
// them. UI builds + executes; this module decides the plan and the verdicts.
import './setup';
import {
  endpointPrivileged,
  classifyResponseOutcome,
  MUTATING_METHODS,
  DEFAULT_DENY_SET,
} from './authz';
import type { Endpoint, Finding, Identity, QaResponse, Severity, Verdict } from './types';

export interface BflaPair {
  endpoint: Endpoint;
  identity: Identity;
  expectation: 'deny';
}

function isPrivilegedIdentity(id: Identity | null | undefined): boolean {
  return !!(id && id.privileged === true);
}

// Derive the BFLA plan: privileged endpoints crossed with non-privileged
// identities (the only pairs where reaching the function is a violation).
export function bflaPlan(
  endpoints: Endpoint[] | null | undefined,
  identities: Identity[] | null | undefined
): BflaPair[] {
  const plan: BflaPair[] = [];
  for (const ep of endpoints || []) {
    if (!endpointPrivileged(ep).privileged) continue;
    for (const id of identities || []) {
      if (isPrivilegedIdentity(id)) continue; // a privileged identity is allowed — not the attacker
      plan.push({ endpoint: ep, identity: id, expectation: 'deny' });
    }
  }
  return plan;
}

// Verdict for a privileged call by a non-privileged identity. Body-aware: a 200
// soft-403 counts as a (correct) denial, not a hole.
export function classifyBfla(
  resp: QaResponse | null | undefined,
  denySet: number[] = DEFAULT_DENY_SET
): Verdict {
  const outcome = classifyResponseOutcome(resp || null, denySet);
  if (outcome === 'allowed') return 'vuln'; // reached a privileged function — BFLA hole
  if (outcome === 'denied') return 'pass';
  return 'inconclusive';
}

export function bflaSeverity(
  method: string | null | undefined,
  verdict: Verdict | string | null | undefined
): Severity | null {
  if (verdict !== 'vuln') return null;
  return MUTATING_METHODS.includes(String(method).toUpperCase()) ? 'critical' : 'high';
}

export function bflaFinding(
  endpoint: Endpoint,
  identity: Identity,
  verdict: Verdict | string
): Finding | null {
  const severity = bflaSeverity(endpoint && endpoint.method, verdict);
  if (!severity) return null;
  return {
    oracle: 'bfla',
    ruleId: 'bfla',
    severity,
    title: 'Broken function-level authorization',
    path: `${endpoint.method} ${endpoint.path}`,
    evidence: `${identity.name || identity.id} (non-privileged) invoked a privileged function`,
    source: 'rule',
  };
}

export interface BflaResultCell {
  endpoint: Endpoint;
  identity: Identity;
  status: number | null;
  verdict: Verdict;
  severity: Severity | null;
  finding: Finding | null;
  response: QaResponse | null;
  error: string | null;
}

/** runBfla 注入的執行器：runner(endpoint, identity) => Promise<resp>。 */
export type BflaRunner = (
  endpoint: Endpoint,
  identity: Identity
) => Promise<QaResponse | null | undefined>;

// Run the derived plan through the injected runner. Streams each finished cell via
// opts.onCell and returns the collected cells + findings. Honors opts.signal.
export async function runBfla(
  endpoints: Endpoint[] | null | undefined,
  identities: Identity[] | null | undefined,
  runner: BflaRunner,
  opts: {
    signal?: AbortSignal | null;
    denySet?: number[];
    onCell?: (cell: BflaResultCell) => void;
  } = {}
): Promise<{ results: BflaResultCell[]; findings: Finding[] }> {
  const { signal, onCell } = opts;
  const denySet = opts.denySet || DEFAULT_DENY_SET;
  const results: BflaResultCell[] = [];
  const findings: Finding[] = [];
  for (const pair of bflaPlan(endpoints, identities)) {
    if (signal && signal.aborted) return { results, findings };
    let cell: BflaResultCell;
    try {
      const resp = await runner(pair.endpoint, pair.identity);
      const status = resp && typeof resp.status === 'number' ? resp.status : null;
      const verdict = classifyBfla(resp || null, denySet);
      const severity = bflaSeverity(pair.endpoint.method, verdict);
      const finding = bflaFinding(pair.endpoint, pair.identity, verdict);
      cell = {
        endpoint: pair.endpoint,
        identity: pair.identity,
        status,
        verdict,
        severity,
        finding,
        response: resp || null,
        error: null,
      };
    } catch (e: any) {
      cell = {
        endpoint: pair.endpoint,
        identity: pair.identity,
        status: null,
        verdict: 'inconclusive',
        severity: null,
        finding: null,
        response: null,
        error: String((e && e.message) || e),
      };
    }
    results.push(cell);
    if (cell.finding) findings.push(cell.finding);
    if (onCell) onCell(cell);
  }
  return { results, findings };
}

File: src-tauri/core/src/security/bfla.rs
//! Port of src/qa/bfla.ts — BFLA (Broken Function-Level Authorization) pure logic.
//! OWASP API5: every privileged endpoint × every non-privileged identity, expect deny.
//! Reuses authz::{endpoint_privileged, classify_response_outcome, is_mutating_method,
//! DEFAULT_DENY_SET, Outcome, Verdict}. The runner lives in this same file (mirrors bola.rs).
use crate::security::authz::{
    classify_response_outcome, endpoint_privileged, is_mutating_method, Outcome, Verdict,
};
use crate::security::finding::{EngineId, Finding, Severity};
use serde_json::Value;

// ── Abstract descriptors (pure helper is callable by runner and fixture bridge) ──

pub struct BflaEndpoint {
    pub method: String,
    pub path: String, // URL path, for privileged-detection only
    pub privileged: Option<bool>,
}

pub struct BflaIdentity {
    pub id: String,
    pub privileged: bool,
}

pub struct BflaPair {
    pub endpoint_index: usize,
    pub identity_index: usize,
}

// ── Pure helpers (TS-golden-fixtured) ─────────────────────────────────────────

/// bflaPlan (bfla.ts:20-31): privileged endpoints × non-privileged identities, endpoint-major.
pub fn bfla_plan(endpoints: &[BflaEndpoint], identities: &[BflaIdentity]) -> Vec<BflaPair> {
    let mut pairs = Vec::new();
    for (ei, ep) in endpoints.iter().enumerate() {
        if !endpoint_privileged(ep.privileged, &ep.method, &ep.path) {
            continue;
        }
        for (ii, id) in identities.iter().enumerate() {
            if id.privileged {
                continue; // privileged identity is not the attacker
            }
            pairs.push(BflaPair {
                endpoint_index: ei,
                identity_index: ii,
            });
        }
    }
    pairs
}

/// classifyBfla (bfla.ts:35-40): Allowed → Vuln, Denied → Pass, Other → Inconclusive.
pub fn classify_bfla(status: Option<i64>, body: &Value, deny_set: &[i64]) -> Verdict {
    match classify_response_outcome(status, body, deny_set) {
        Outcome::Allowed => Verdict::Vuln,
        Outcome::Denied => Verdict::Pass,
        Outcome::Other => Verdict::Inconclusive,
    }
}

/// bflaSeverity (bfla.ts:42-45): Vuln + mutating → Critical, Vuln + read → High, else None.
pub fn bfla_severity(method: &str, verdict: Verdict) -> Option<Severity> {
    if verdict != Verdict::Vuln {
        return None;
    }
    Some(if is_mutating_method(method) {
        Severity::Critical
    } else {
        Severity::High
    })
}

/// bflaFinding: build a Finding for a BFLA violation. Returns None for non-vuln verdicts.
/// NOTE: evidence uses identity_id directly (Rust Identity has no `name` field; TS used
/// `identity.name || identity.id`). Callers MUST uppercase method ONCE before calling.
pub fn bfla_finding(
    method: &str,
    request_id: &str,
    identity_id: &str,
    verdict: Verdict,
) -> Option<Finding> {
    let severity = bfla_severity(method, verdict)?;
    Some(Finding {
        engine: EngineId::Bfla,
        severity,
        rule_id: "bfla".into(),
        oracle: "bfla".into(),
        title: "Broken function-level authorization".into(),
        path: format!("{} {}", method, request_id),
        evidence: format!(
            "{} (non-privileged) invoked a privileged function",
            identity_id
        ),
        method: Some(method.into()),
        endpoint: Some(request_id.into()),
        identity: Some(identity_id.into()),
    })
}

// ── Runner ────────────────────────────────────────────────────────────────────

use crate::config::{Config, Identity, Request};
use crate::security::finding::EngineError;
use crate::security::runner::run_security_step;

/// run_bfla: mirrors run_bola's structure. Returns (vec![], vec![]) when no security.bfla block.
pub async fn run_bfla(cfg: &Config, env: Option<&str>) -> (Vec<Finding>, Vec<EngineError>) {
    let bcfg = match cfg.security.as_ref().and_then(|s| s.bfla.as_ref()) {
        Some(b) => b,
        None => return (vec![], vec![]),
    };
    // Empty `denySet: []` is honored as-is (user says no status is a hard-deny), like matrix; an
    // ABSENT denySet already defaulted to [401,403,404] via serde at config load.
    let deny_set = bcfg.deny_set.clone();

    // Build filtered request list (bfla.endpoints non-empty ⇒ filter; else all).
    let reqs: Vec<&Request> = if bcfg.endpoints.is_empty() {
        cfg.requests.iter().collect()
    } else {
        cfg.requests
            .iter()
            .filter(|r| bcfg.endpoints.contains(&r.id))
            .collect()
    };

    // Build parallel BflaEndpoint descriptors. `path` is used ONLY for privileged-detection
    // (admin-path tokens) — parse the RAW url exactly like run_matrix (runner.rs): a templated
    // url fails Url::parse and falls back to the raw string, which still carries the path tokens.
    // No {{var}} substitution here — run_security_step does its own substitution via build_request.
    let endpoints: Vec<BflaEndpoint> = reqs
        .iter()
        .map(|req| {
            let path = reqwest::Url::parse(&req.url)
                .map(|u| u.path().to_string())
                .unwrap_or_else(|_| req.url.clone());
            BflaEndpoint {
                method: req.method.clone(),
                path,
                privileged: req.privileged,
            }
        })
        .collect();

    // All identities.
    let identities_ref: Vec<&Identity> = cfg.identities.iter().collect();
    let bfla_identities: Vec<BflaIdentity> = identities_ref
        .iter()
        .map(|i| BflaIdentity {
            id: i.id.clone(),
            privileged: i.privileged,
        })
        .collect();

    let plan = bfla_plan(&endpoints, &bfla_identities);

    let mut findings: Vec<Finding> = Vec::new();
    let mut errors: Vec<EngineError> = Vec::new();

    for pair in &plan {
        let req = reqs[pair.endpoint_index];
        let identity = identities_ref[pair.identity_index];
        let req_id = &req.id;
        let method = req.method.to_uppercase(); // uppercase ONCE

        let step = run_security_step(cfg, req, identity, env).await;
        if !step.success {
            errors.push(EngineError {
                engine: EngineId::Bfla,
                endpoint: Some(req_id.clone()),
                identity: Some(identity.id.clone()),
                // method+req_id only; step.error may carry the resolved URL → redacted at scan output.
                message: format!(
                    "{} {}: {}",
                    method,
                    req_id,
                    step.error.as_deref().unwrap_or("request failed")
                ),
            });
            continue;
        }

        let verdict = classify_bfla(Some(step.status), &step.body, &deny_set);
        if let Some(finding) = bfla_finding(&method, req_id, &identity.id, verdict) {
            findings.push(finding);
        }
    }

    (findings, errors)
}

// ── Unit tests ────────────────────────────────────────────────────────────────

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    fn ep(method: &str, path: &str, privileged: Option<bool>) -> BflaEndpoint {
        BflaEndpoint {
            method: method.into(),
            path: path.into(),
            privileged,
        }
    }
    fn id(id: &str, privileged: bool) -> BflaIdentity {
        BflaIdentity {
            id: id.into(),
            privileged,
        }
    }

    #[test]
    fn bfla_plan_crosses_priv_ep_with_nonpriv_id() {
        let eps = vec![ep("DELETE", "/admin/x", None), ep("GET", "/u", None)];
        let ids = vec![id("admin", true), id("anon", false)];
        let plan = bfla_plan(&eps, &ids);
        assert_eq!(plan.len(), 1);
        assert_eq!(plan[0].endpoint_index, 0); // DELETE /admin/x (priv via heuristic)
        assert_eq!(plan[0].identity_index, 1); // anon (non-priv)
    }

    #[test]
    fn bfla_plan_explicit_privileged_override() {
        // explicit privileged:true on a GET (non-mutating) => still included
        let eps = vec![ep("GET", "/u", Some(true))];
        let ids = vec![id("anon", false)];
        let plan = bfla_plan(&eps, &ids);
        assert_eq!(plan.len(), 1);
        // explicit privileged:false on a DELETE => excluded
        let eps2 = vec![ep("DELETE", "/admin/x", Some(false))];
        let plan2 = bfla_plan(&eps2, &ids);
        assert_eq!(plan2.len(), 0, "explicit false overrides DELETE heuristic");
    }

    #[test]
    fn classify_bfla_cases() {
        let deny = &[401i64, 403, 404];
        assert_eq!(classify_bfla(Some(200), &json!({}), deny), Verdict::Vuln);
        assert_eq!(classify_bfla(Some(403), &json!({}), deny), Verdict::Pass);
        assert_eq!(
            classify_bfla(Some(500), &json!({}), deny),
            Verdict::Inconclusive
        );
        assert_eq!(classify_bfla(None, &json!({}), deny), Verdict::Inconclusive);
        // soft-deny: 200 with error body => denied => pass
        assert_eq!(
            classify_bfla(Some(200), &json!({"error":"Access denied"}), deny),
            Verdict::Pass
        );
    }

    #[test]
    fn bfla_severity_cases() {
        assert_eq!(
            bfla_severity("DELETE", Verdict::Vuln),
            Some(Severity::Critical)
        );
        assert_eq!(
            bfla_severity("POST", Verdict::Vuln),
            Some(Severity::Critical)
        );
        assert_eq!(
            bfla_severity("PUT", Verdict::Vuln),
            Some(Severity::Critical)
        );
        assert_eq!(
            bfla_severity("PATCH", Verdict::Vuln),
            Some(Severity::Critical)
        );
        assert_eq!(bfla_severity("GET", Verdict::Vuln), Some(Severity::High));
        assert_eq!(bfla_severity("get", Verdict::Vuln), Some(Severity::High)); // case-insensitive via is_mutating_method
        assert_eq!(bfla_severity("DELETE", Verdict::Pass), None);
        assert_eq!(bfla_severity("GET", Verdict::Inconclusive), None);
    }

    #[test]
    fn bfla_finding_shape() {
        let f = bfla_finding("DELETE", "deleteAdmin", "anon", Verdict::Vuln).unwrap();
        assert_eq!(f.engine, EngineId::Bfla);
        assert_eq!(f.severity, Severity::Critical);
        assert_eq!(f.rule_id, "bfla");
        assert_eq!(f.oracle, "bfla");
        assert_eq!(f.title, "Broken function-level authorization");
        assert_eq!(f.path, "DELETE deleteAdmin");
        assert_eq!(
            f.evidence,
            "anon (non-privileged) invoked a privileged function"
        );
        assert_eq!(f.method, Some("DELETE".into()));
        assert_eq!(f.endpoint, Some("deleteAdmin".into()));
        assert_eq!(f.identity, Some("anon".into()));
    }

    #[test]
    fn bfla_finding_none_for_non_vuln() {
        assert!(bfla_finding("DELETE", "r", "anon", Verdict::Pass).is_none());
        assert!(bfla_finding("DELETE", "r", "anon", Verdict::Inconclusive).is_none());
    }

    #[test]
    fn bfla_finding_get_is_high() {
        let f = bfla_finding("GET", "getAdmin", "anon", Verdict::Vuln).unwrap();
        assert_eq!(f.severity, Severity::High);
    }
}
